### DISCUSSION TITLE
Implementation of feature/cards_and_boxes_using_bslib

### DISCUSSION THREAD

**Discussion by tinneuro** [United4Surveillance/signal-detection-tool#289] [Show and tell]
Hi all,
I just discovered the Discussions Tab on Github. Thought it might be an idea to use it for the branch @aj-thl made. So we can give feedback here now before he makes a PR. We can also make a PR and comment there, whatever....
The discussions forum could still be useful also for other things.
It is the feature/cards_and_boxes_using_bslib branch

**Comment by aj-thl** [comment]
Hi and thanks Ann Christin for opening this thread! This discussion is mainly about the issue #183 

As I said yesterday, I have done some testing with the bslib package (branch: feature/cards_and_boxes_using_bslib). In short, boxes are now implemented by using the bslib and, I have added cards feature for the max three stratification plots. However, to make everything work as intended, I was forced to change the shiny::fluidPage() with the bslib::page_fluid() in the app_ui.R. Another solution would be to stick with the shiny::fluidPage() and add bslib::bs_theme() but I think the layout would be basically the same.

Anyway, I noticed that when bslib::page_fluid() is used the layout of the tabset titles is different when running the app in Rstudio. However, if one runs the app in web browser the layout of the tabset titles is quite same as they are in the current version of the app. I’m not entirely sure why this happens. I would guess that the viewer in Rstudio might not fully support all functionalities included in the bslib. 

It would be nice to hear your valuable thoughts and feedback before I proceed any further with this. I would especially appreciate if you could give feedback about the layout if you like or not. All suggestions for improvement are also very welcome! @tinneuro @tomasmartbert @alchalu @evolutionv2

**Comment by tinneuro** [comment]
Hi @aj-thl,

Thanks a lot for trying bslib out! I like the new layout, in particular I really like the design of the boxes for the stratified plots. 
For me there is no issue with the tabs looking differently in Rstudio, they look good. Maybe this is a version issue? I am using R 4.2.2 and bslib 0.7.0 (which is also the minimal required version because of value_box_theme()). What did you use?
In general I think it would be nice to move to bslib and integrate your branch. :)
I have some comments about the boxes but also thought about the general design of the page which we could now improve.

- I think the boxes are too big thus take too much space. Can we reduce them in length
- I think we could reduce to two boxes and get rid of the Outbreak Detection Algorithm box but opinions of others very welcome. We could also think of adding the outbreak detection algorithm info somewhere else( dont know where yet)
- I think we can also improve now on my first header version "Visualisations and/or tables showing the number of cases in the last 12 weeks with alarms from Signal Detection for the selected strata age_group, sex, county." I would just completely delete it. Instead it would be nice to have headers for each of the cards with i.e. "Distribution by age_group, CW 13-17 2024" Thus I am also pro adding the Calenderweek or the explicit date instead of the number of weeks.
- I don't like that the interesting information about alarms per stratum are put in a small box which is then scrollable. 
My suggestion is for the second box: CW: x1 then below Number of cases: x2 then below Number of alarms: x3. x2 would be number of cases unstratified. CW: calendarweek can be either like CW 13-17 2024 or even the exact dates. I think this is much nicer than putting the number of weeks. Number of alarms would be the one we are currently displaying in the box above the stratified alarms. With this we could increase the box for the stratified alarms to the full height. Here I would add a line on top with national level: x4 x4 being the number of alarms on the unstratified timeseries. The Question is whether we should call that national level, in most cases it will be but they could also upload a subset of data which is not the national level. Thus we could also put unstratified or so. Will add another comment with how we could call it. ideas welcome.
- I would like to move the "Toggle number of alarms on / off on stratification graphs" button below the cards. In addition I want to rename it to "Show number of alarms" and then rather have a apple like switch button instead
https://dreamrs.github.io/shinyWidgets/
- Minor comment: you then need to still add bslib (0.7.0) to the DESCRIPTION Depends file

Thanks a lot!

**Reply by aj-thl** [reply]
Hi @tinneuro! 

Thanks a lot for your valuable thoughts! 

It is good to hear that you are not having issues with the layout of the tabs. I am using R 4.3.1 and I noticed that I had an older version of bslib (0.6.1). Anyway, I updated the bslib and the layout of the titles of the tabs did not change for me. Maybe we get some kind of conclusion if the others will test this too 😄 Personally I'm ok with the layout. It bothers me a bit that the titles of the tabs are quite close to each other but I can live with that. 

But anyway, I agree that the boxes are little too big and I will change that. I also like your suggestions about reorganizing the boxes and make adjustments to the titles. I will definitely do that. 

I will also take a look that switch button that you mentioned and add bslib (0.7.0) to the DESCPITION file. 

Thanks again for your feedback, I appreciate it!

**Comment by tinneuro** [comment]
Great! In addition I added a function called pretty_variable_names() to the levels.R in my new PR #292.
If you want and have time you could work on improving that these are also used to show alarms by strata and in the titles of the stratified plots.

**Comment by tinneuro** [comment]
Hi @aj-thl ,

are you working again on the improvements? Just let us know as soon as you implemented it and then we can do another review round with everyone. Thanks!

**Reply by aj-thl** [reply]
Hi! Yes, sorry for the delay. I have been busy with another project but I try to finish the improvements during this week (most likely on Friday) so that we could have another review round during the next week. Thanks for your patience.

**Reply by tinneuro** [reply]
No worries just wanted to get informed about the status. Great that sounds good. Thanks!

**Comment by aj-thl** [comment]
Ok, so I did few things today. 

Firstly, the number of boxes are now reduced to two. I'm not sure did you really wanted that the number of alarms (national level) is present in both boxes. Maybe we could include this information only to the "number of alarms" box? 

Secondly, I replaced the old "Toggle number of alarms on / off on stratification graphs" button with a new one as you suggested. For some reason now the whole graph disappears if you turn off alarms. After a quick look I could not find a clear reason for this. Maybe you others will see better with fresh eyes and spot the mistake. 

I still need to introduce new tidy category names for the headings of the cards and probably tidy some code also because I did this in a hurry. 

But anyway, feedback during the next week would be highly appreciated!

**Reply by tinneuro** [reply]
Great, thank you @aj-thl. I will have a look at it this week.

**Comment by aj-thl** [comment]
Hi! 

I made one bug fix this morning and also introduced the pretty variable names in cards and boxes. 

Still need to figure out the issue with the alarms on/off button.

**Reply by aj-thl** [reply]
I found the cause of the problem with the button. I will push the updated code by the end of this day.

**Reply by aj-thl** [reply]
The up-to-date code should now be available.

**Comment by alchalu** [comment]
Thank you for this version with bslib! Just a couple of things that I noticed: 

- I really like the cards for the plots! 
- Some of the text in the app looks a bit weird now, especially the title is smushed in the top left corner 
- I think the start end end cw are switched 
- If the period includes the turn of the year it should say something like CW 40, 2022 to CW 15, 2023
- I would prefere a version with more boxes and less info in each box 
- maybe the width of the boxes can adjust dynamically 

I also get an error for the assertion if `toggle_alarms` is logical. It works anyway but by adding `shiny::req(input$alarms_trig)` before the call to `decider_barplot_map_table()` in the signal tab panel, shiny won't be confused anymore 😄

**Comment by tinneuro** [comment]
Hi,
FYI I will give you a summary of what we discussed in the meeting on thursday later today :)

**Comment by tinneuro** [comment]
Hi,

first some minor design details:
- as @alchalu said above the top headline is directly on the top no space being there between the very top of the app and the title. Is there a possibility you can look into this?
- In the discussion we also realised we can just shorten the name because United4Surveillance is there now as emblem. So could you also just make Signal Detection Tool as title?
- We agreed that the boxes are generally too big, can you try out how it looks reducing them in size and we also rather want boxes containing the same topics. (will come to that later)
- The boxes are attached directly below the tab line. Is there a possibility to add some space between the boxes and the tab line?

Content wise: 
- Rename national level to "Unstratified" 
- have the weeks in the same format as in the timeseries plot i.e. W51-2023 - W02-2024 
- the alarms on the now "Unstratified" former "National level" should not be the sum of the alarms of the stratfieid alarms but only the alarms found on the top level. For the test data using age_group and county as strata and farringtonflexible this would result in 0 alarms on the unstratified level
- for the content of the boxes I am not 100% sure anymore what we said. Maybe we can make three boxes in total one box with the info on the calendarweeks and put the algorithm back there as well
- then one box with Unstratified number of cases and alarms
- one with the stratified number of alarms

I unfortunatley dont have my notes right now with me but I think we can go with that for now :) But others are very invited to comment on how exactly the content of the boxes should be

**Reply by aj-thl** [reply]
Hi! Thank you and @alchalu for your feedback. I agree with you on almost everything. 

I start my summer holiday on 8.7. and will be away for 5 weeks, so I'll try to finish this before my holiday.

Thanks again, this was very helpful 🙂

**Comment by tinneuro** [comment]
I found the following note:
"three diffrerent boxes one for weeks, disease, method and one for summary statistic, one for number of alarms"
For summary statistic we only have number of cases right now. Number of alarms could contain two sections one with unstratified and stratified alarms or we put the unstratified alarms to the number of cases which are also unstratified.

**Comment by aj-thl** [comment]
Ok, I made a few changes based on your feedback. There are now three boxes: 

1. Method, disease and period
2. Number of unstratified cases and alarms
3. Number of stratified alarms

The number of unstratified alarms (and cases) should be now calculated correctly. 

I tried to reduce the size of the boxes as small as I could, so that they don't become scrollable (should also apply when using 3 stratas and the signal detection period includes several years). To achieve this, I needed to adjust the font size, margins etc. by using CSS code. Maybe it’s not perfect, but I’m quite happy with the result. I also added some space around the title of the app and above the boxes. 

The title of the app is also now changed as requested. 

As @alchalu mentioned, the shiny gave a warning for the assertion toggle_alarms is not logical. It seems that the `shinyWidgets::materialSwitch` returns `NULL` instead of `FALSE` when toggled off and this was the cause of the problem. This should now also be resolved.

Next, I will make a pull request. I would be grateful if someone could look into it early next week as I will be on holiday for 5 weeks from Monday 8.7. ping @tinneuro
